File: frontend/src/types/simulation.ts
export type SolverType = 'euler' | 'rk4' | 'merson'

export interface SimulationConfig {
  solver: SolverType
  startTime: number
  stopTime: number
  stepSize: number
  maxStep?: number
  minStep?: number
  relativeTolerance?: number
  absoluteTolerance?: number
  maxResultPoints?: number
}

export type SimulationStatus =
  | 'idle'
  | 'compiling'
  | 'running'
  | 'paused'
  | 'completed'
  | 'error'

export interface SimulationState {
  status: SimulationStatus
  currentTime: number
  progress: number
  error?: string
}

export interface SignalData {
  blockId: string
  portId: string
  name: string
  times: number[]
  values: number[] | number[][]  // Single array for single-input, array of arrays for multi-input
  inputNames?: string[]  // Names for each input trace (for legend)
  numInputs?: number  // Number of inputs (1 for single-input scopes)
  // 3D Scope specific fields
  x?: number[]
  y?: number[]
  z?: number[]
  is3D?: boolean
}

export type AnalysisType = 'bode' | 'nyquist' | 'pzmap' | 'stepinfo'

export interface AnalysisData {
  analysisType: AnalysisType
  name?: string  // Block name for display
  // Bode-specific
  frequencies?: number[]
  magnitude_db?: number[]
  phase_deg?: number[]
  gain_margin?: number | null
  phase_margin?: number | null
  gain_crossover_freq?: number | null
  phase_crossover_freq?: number | null
  // Nyquist-specific
  real?: number[]
  imag?: number[]
  encirclements?: number
  // Pole-Zero specific
  poles?: [number, number][]
  zeros?: [number, number][]
  is_stable?: boolean
  dominant_pole?: [number, number] | null
  // Step response specific
  times?: number[]
  response?: number[]
  rise_time?: number | null
  settling_time?: number | null
  overshoot_percent?: number | null
  peak_time?: number | null
  peak_value?: number | null
  steady_state_value?: number | null
}

export interface SimulationResults {
  signals: SignalData[]
  analyses?: Record<string, AnalysisData>  // Analysis block data keyed by block ID
  statistics: {
    totalSteps: number
    executionTime: number
    finalTime: number
    decimationFactors?: Record<string, number>
  }
}

File: backend/src/models/simulation.py
"""Simulation-related Pydantic models."""

from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class SolverType(str, Enum):
    """Available ODE solvers."""

    EULER = "euler"
    RK4 = "rk4"
    MERSON = "merson"


class SimulationConfig(BaseModel):
    """Simulation configuration."""
    model_config = ConfigDict(populate_by_name=True)

    solver: SolverType = SolverType.RK4
    start_time: float = Field(default=0.0, alias="startTime")
    stop_time: float = Field(default=10.0, alias="stopTime")
    step_size: float = Field(default=0.01, alias="stepSize")
    max_step: float | None = Field(default=None, alias="maxStep")
    min_step: float | None = Field(default=None, alias="minStep")
    relative_tolerance: float = Field(default=1e-3, alias="relativeTolerance")
    absolute_tolerance: float = Field(default=1e-6, alias="absoluteTolerance")


class SimulationStatus(str, Enum):
    """Simulation execution status."""

    IDLE = "idle"
    COMPILING = "compiling"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    ERROR = "error"


class SignalData(BaseModel):
    """Time series data for a signal."""
    model_config = ConfigDict(populate_by_name=True)

    block_id: str = Field(alias="blockId")
    port_id: str = Field(alias="portId")
    name: str
    times: list[float]
    values: list[float]


class SimulationStatistics(BaseModel):
    """Simulation execution statistics."""
    model_config = ConfigDict(populate_by_name=True)

    total_steps: int = Field(alias="totalSteps")
    execution_time: float = Field(alias="executionTime")
    final_time: float = Field(alias="finalTime")


class SimulationResults(BaseModel):
    """Complete simulation results."""

    signals: list[SignalData]
    statistics: SimulationStatistics
